AdblockPlusSafari/AdblockPlusExtras.m

```diff
@@ -0,0 +1,269 @@
+    // Force content blocker to load newer version of filterlist
+    [self reloadContentBlockerWithCompletion:nil];
+  }
+}
+
+#pragma mark -
+
+- (void)setEnabled:(BOOL)enabled reload:(BOOL)reload
+{
+  self.enabled = enabled;
+
+  if (reload) {
+    [self reloadContentBlockerWithCompletion:nil];
+  }
+}
+
+- (void)setAcceptableAdsEnabled:(BOOL)enabled reload:(BOOL)reload
+{
+  self.acceptableAdsEnabled = enabled;
+
+  if (reload) {
+    [self reloadContentBlockerWithCompletion:nil];
+  }
+}
+
+- (void)reloadContentBlockerWithCompletion:(void(^__nullable)(NSError * __nullable error))completion;
+{
+  __weak __typeof(self) wSelf = self;
+  wSelf.reloading = YES;
+  [SFContentBlockerManager reloadContentBlockerWithIdentifier:self.contentBlockerIdentifier completionHandler:^(NSError *error) {
+    NSLog(@"%@", error);
+    dispatch_async(dispatch_get_main_queue(), ^{
+      wSelf.reloading = NO;
+      [wSelf checkActivatedFlag];
+      if (completion) {
+        completion(error);
+      }
+    });
+  }];
+}
+
+- (void)checkActivatedFlag
+{
+  BOOL activated = [self.adblockPlusDetails boolForKey:AdblockPlusActivated];
+  if (self.activated != activated) {
+    self.activated = activated;
+  }
+}
+
+- (void)updateFilterlists
+{
+  NSMutableDictionary *filterlists = [self.filterlists mutableCopy];
+  for (NSString *filterlistName in self.filterlists) {
+    NSURL *url = [NSURL URLWithString:filterlistName];
+
+    NSURLSessionTask *task = [self.backgroundSession downloadTaskWithURL:url];
+
+    NSMutableDictionary *filterlist = [filterlists[filterlistName] mutableCopy];
+    filterlist[@"updating"] = @YES;
+    filterlist[@"taskIdentifier"] = @(task.taskIdentifier);
+    filterlists[filterlistName] = filterlist;
+
+    [self.downloadTasks[filterlistName] cancel];
+    // Store key to task cache
+    self.downloadTasks[filterlistName] = task;
+
+    [task resume];
+  }
+  self.filterlists = filterlists;
+}
+
+#pragma mark - NSURLSessionDownloadDelegate
+
+- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error
+{
+  NSString *filterlistName = task.originalRequest.URL.absoluteString;
+  NSDictionary *filterlist = self.filterlists[filterlistName];
+
+  if ([filterlist[@"taskIdentifier"] unsignedIntegerValue] == task.taskIdentifier && [filterlist[@"updating"] boolValue]) {
+
+    NSMutableDictionary *mutableFilterlist = [filterlist mutableCopy];
+    mutableFilterlist[@"updating"] = @NO;
+    [mutableFilterlist removeObjectForKey:@"taskIdentifier"];
+
+    NSMutableDictionary *mutableFilterlists = [self.filterlists mutableCopy];
+    mutableFilterlists[filterlistName] = mutableFilterlist;
+    self.filterlists = mutableFilterlists;
+
+    // Remove key from task cache
+    [self.downloadTasks removeObjectForKey:filterlistName];
+  }
+}
+
+- (void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didFinishDownloadingToURL:(NSURL *)location
```

Comment:
I'm not sure how this is actually invoked, but we should make sure that we wait one minute before attempting the first background download after the app starts. I'll add that to the issue.